crates/rome_js_analyze/src/analyzers/nursery/no_confusing_void_type.rs

```diff
@@ -0,0 +1,157 @@
+    /// > If you think you need this then you probably want the undefined type instead.
+    ///
+    /// ## Examples
+    /// ### Invalid
+    ///
+    /// ```ts,expect_diagnostic
+    /// type PossibleValues = number | void;
+    /// type MorePossibleValues = string | ((number & any) | (string | void));
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// function logSomething(thing: void) {}
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// interface Interface {
+    ///     prop: void;
+    /// }
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// let foo: void;
+    /// let bar = 1 as unknown as void;
+    /// let baz = 1 as unknown as void | string;
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```ts
+    /// function foo(): void {};
+    /// function doSomething(this: void) {}
+    /// function printArg<T = void>(arg: T) {}
+    /// logAndReturn<void>(undefined);
+    /// let voidPromise: Promise<void> = new Promise<void>(() => { });
+    /// let voidMap: Map<string, void> = new Map<string, void>();
+    /// ```
+    ///
+    pub(crate) NoConfusingVoidType {
+        version: "1.0.0",
+        name: "noConfusingVoidType",
+        recommended: false,
+    }
+}
+
+type Language = <AnyTsType as AstNode>::Language;
+
+// We only focus on union type
+pub enum VoidTypeIn {
+    Union,
+    Unknown,
+}
+
+impl Rule for NoConfusingVoidType {
+    type Query = Ast<AnyTsType>;
+    type State = VoidTypeIn;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+
+        if let AnyTsType::TsVoidType(node) = node {
+            let result = node_in(node.syntax());
+            return result;
+        }
+
+        None
+    }
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+        return Some(RuleDiagnostic::new(
+            rule_category!(),
+            node.range(),
+            markup! {{match_message(state)}},
+        ));
+    }
+}
+
+fn node_in(node: &SyntaxNode<Language>) -> Option<VoidTypeIn> {
+    let parent = node.parent()?;
+
+    // (string | void)
+    if TsParenthesizedType::can_cast(parent.kind()) {
+        return node_in(&parent);
+    }
+
+    // string | void
+    if TsUnionTypeVariantList::can_cast(parent.kind()) {
+        return node_in(&parent);
+    }
+
+    // string & void
+    if TsIntersectionTypeElementList::can_cast(parent.kind()) {
+        return node_in(&parent);
+    }
+
+    // arg: void
+    if TsTypeAnnotation::can_cast(parent.kind()) {
+        return node_in(&parent);
+    }
+
+    // fn<T = void>() {}
+    // T = void
+    if TsDefaultTypeClause::can_cast(parent.kind()) {
+        return node_in(&parent);
+    }
+
+    // string | void or string & void
+    if let Some(n) = parent.cast_ref::<AnyTsType>() {
```

Comment:
If you move this at the end of the end (or in the match I talked), you can avoid cloning by directly calling `cast` and consuming `parent` (because `parent` is no longer used at the end)